### PR TITLE
Support correlated scalar subqueries (v3)

### PR DESCRIPTION
**Pull request by kokosing**
Support correlated count scalar subqueries

A scalar subquery is a simple query which aggregates all the rows into a
single group (empty GROUP BY clause). For example:

SELECT min(a) FROM t WHERE b > 0
This commit add support for correlated count scalar subqueries which
have WHERE cluase of conjunction of equality comparison. For example:

SELECT (SELECT min(a) FROM t1(a,c) WHERE b = 0 AND c = d) FROM t2(b,d)
Logically such query is rewritten into left outer join where correlation
used in eqality comparisons (from WHERE cluase) are used as join
condition.

### CODE REVIEW

**Comment by kokosing** [comment]
Superse

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 1]
```diff
@@ -308,6 +308,20 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
                     InPredicate inPredicate = Iterables.getLast(inPredicates);
                     throwNotSupportedException(inPredicate, "IN with subquery predicate in join condition");
                 }
+
+                subqueryPlanner.collectExistsSubqueries(complexExpression, node).stream()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: add tests

**Comment by kokosing** [comment]
@martint Supersedes #5824 and depends on #5983

**Comment by martint** [comment]
I merged some of the refactoring commits.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 2]
```diff
@@ -407,4 +407,10 @@ public PlanNode visitValues(ValuesNode node, RewriteContext<Void> context)
                     rewrittenRows);
         }
     }
+
+    public boolean isCorrelated(Query subquery)
+    {
+        PlanNode subqueryPlan = createRelationPlan(subquery).getRoot();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: `createRelationPlan` has side-effects, so this is probably not a very good idea to do in a method that appears to be side effect-free.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeProjections.java [hunk 3]
```diff
@@ -60,21 +60,30 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
...
 
-                return new ProjectNode(node.getId(), ((ProjectNode) source).getSource(), projections.build());
+                    return new ProjectNode(node.getId(), sourceProject.getSource(), projections.build());
+                }
             }
-
             return replaceChildren(node, ImmutableList.of(source));
         }
+
+        private static boolean isDeterministic(ProjectNode second)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeProjections.java]: Name the argument `node` instead of `second`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeProjections.java [hunk 4]
```diff
@@ -60,21 +60,30 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
...
+        {
+            return second.getAssignments().values().stream().allMatch(DeterminismEvaluator::isDeterministic);
+        }
+
+        private static boolean doNotUseTryExpression(ProjectNode first)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeProjections.java]: Name the argument `node` instead of `first`

Also, I'd rename the method to `containsTry` and change the usage to do `!containsTry(node)`

**Comment by martint** [comment]
I also merged 4911aaf601fe7a351ec6d4d5ebd35d8abd334216 after applying the comments I made above.

File: presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class EnforceUniqueColumnsOperator
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java]: I'd call this AssignUniqueIdOperator

File: presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new UuidOperatorFactory(operatorId, planNodeId, types);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private boolean finishing;
+    private final int taskId;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java]: Note that if rows from two different stages are merged together (e.g., via a union), task ids may overlap and the values won't be unique anymore. Also, a broadcast after this operator will break the uniqueness.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+import com.facebook.presto.sql.planner.Symbol;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class EnforceUniqueColumns
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java]: I'd call this AssignUniqueId

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        extends PlanNode
+{
+    private final PlanNode source;
+    private final Symbol mostSigBitsSymbol;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java]: I'd call this `highBits` and the one below `lowBits`.

BTW, do we really need two fields? We could probably just use 64 bits. Since we're not generating them randomly, it should be sufficient.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeUtils.findFirstPlanNodeMatching;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeUtils.removeFirstPlanNodesMatching;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeUtils.replaceAllPlanNodesMatching;
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: Can you add some comments describing what the shape of the before and after plan look like?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 10]
```diff
@@ -375,7 +375,8 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                     if (joinEqualityExpression(node.getLeft().getOutputSymbols()).test(conjunct)) {
                         ComparisonExpression equality = (ComparisonExpression) conjunct;
 
-                        boolean alignedComparison = Iterables.all(DependencyExtractor.extractUnique(equality.getLeft()), in(node.getLeft().getOutputSymbols()));
+                        boolean alignedComparison = Iterables.all(DependencyExtractor.extractUnique(equality.getLeft()), in(node.getLeft().getOutputSymbols()))
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: How is this change related to this commit? Or was this a fix for another issue?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            ApplyNode rewrittenNode = (ApplyNode) context.defaultRewrite(node, context.get());
+            if (!rewrittenNode.getCorrelation().isEmpty()) {
+                Optional<AggregationNode> scalarAggregation = findFirstPlanNodeMatching(
+                        rewrittenNode.getSubquery(),
+                        AggregationNode.class::isInstance,
+                        Predicates.<PlanNode>isInstance(ProjectNode.class).or(EnforceSingleRowNode.class::isInstance));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: I think this would be more readable as:

```java
e -> e instanceof ProjectNode || e instanceof EnforceSingleRowNode
```

or introduce a helper method to construct a predicate like:

```java
Predicates.instanceOfAny(ProjectNode.class, EnfoceSingleRowNode.class)
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java [hunk 12]
```diff
@@ -66,5 +72,22 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
             }
             return rewrittenNode;
         }
+
+        private boolean hasScalarSubquery(ApplyNode node)
+        {
+            return node.getSubquery() instanceof EnforceSingleRowNode;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: Technically, a ValuesNode with a single row would qualify, no?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java [hunk 13]
```diff
@@ -66,5 +72,22 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
...
+        }
+
+        private boolean hasScalarAggregationSubquery(ApplyNode node)
+        {
+            return findFirstPlanNodeMatching(
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: This is hard to read due to arguments being lined up with the chained calls. Unfortunately, there's no way to indent them differently in IntelliJ. 

I'd introduce a temporary variable to make it clear what's going on:

```java
Optional<PlanNode> aggregation = findFirstPlanNodeMatching(
        node.getSubquery(),
        AggregationNode.class::isInstance,
        Predicates.<PlanNode>isInstance(ProjectNode.class).or(LimitNode.class::isInstance));

return aggregation
        .map(AggregationNode.class::cast)
        .map(AggregationNode::getGroupBy)
        .map(List::isEmpty)
        .orElse(false);
```

Alternatively, we could introduce a builder-style matcher and rewrite it like:

```java
return matcher(node.getSubquery())
    .where(AggregationNode.class::isInstance)
    .skip(instanceOfAny(ProjectNode.class, LimitNode.class)
    .find()
    .map(AggregationNode.class::cast)
    .map(AggregationNode.getGroupBy)
    .map(List::isEmpty)
    .orElse(false)
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoApplyNodeLeftChecker.java [hunk 14]
```diff
@@ -39,7 +39,7 @@ public Object visitApply(ApplyNode node, Object context)
                     throw new IllegalArgumentException("Not supported subquery");
                 }
                 else {
-                    throw new IllegalArgumentException("Correlated subquery is not yet supported");
+                    throw new IllegalArgumentException("Not supported correlated subquery");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoApplyNodeLeftChecker.java]: Why? The new phrasing is weird, too. I'd probably say "Unsupported correlated subquery type"

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 15]
```diff
@@ -6038,9 +6038,110 @@ public void testCorrelatedScalarSubqueries()
...
+                        "FROM orders o " +
+                        "GROUP BY o.orderkey " +
+                        "HAVING 40000 < (SELECT avg(i.orderkey) FROM orders i WHERE o.orderkey < i.orderkey AND i.orderkey % 10000 = 0)" +
+                        "ORDER BY o.orderkey LIMIT 1",
+                "VALUES (68690.22, 20000)"); // h2 is slow
+        assertQuery(
+                "SELECT max(o.totalprice), o.orderkey FROM orders o " +
+                        "GROUP BY o.orderkey, (SELECT avg(i.orderkey) FROM orders i WHERE o.orderkey < i.orderkey AND i.orderkey % 10000 = 0)" +
+                        "ORDER BY o.orderkey LIMIT 1",
+                "VALUES (172799.49, 1)"); // h2 is slow
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: This is brittle. The result of floating point arithmetic is sensitive to the order the operations are applied. For example, this may produce different results depending on whether the query executes in a single core or is parallelized across multiple cores or machines. We may want to choose another aggregation function to use, instead.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 16]
```diff
@@ -6091,24 +6192,150 @@ public void testCorrelatedExistsSubqueriesWithPrunedCorrelationSymbols()
...
+                "      l2.orderkey = l1.orderkey\n" +
+                "  )\n" +
+                "  AND NOT EXISTS(\n" +
+                "    SELECT *\n" +
+                "    FROM\n" +
+                "      orders l3\n" +
+                "    WHERE\n" +
+                "      l3.orderkey = l1.orderkey\n" +
+                "  )\n",
+                "VALUES 0"); // EXISTS predicates are contradict
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: type: contradictory

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java (hunk 9 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: Also, I'd call this class `TransformCorrelatedScalarAggregationToJoin` for brevity.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+    }
+
+    private static class ScalarAggregationSourceRewriter
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: I don't think we need a class for this. There's no state that needs to be maintained, so it could just be a bunch of static methods in the enclosing class.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        private ScalarAggregationSourceRewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        public Optional<Result> rewrite(AggregationNode scalarAggregation, List<Symbol> correlation)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: This method could be made a bit more readable by untangling some of the things it's doing. Something like:

```java
public Optional<Result> rewrite(AggregationNode aggregation, List<Symbol> boundVariables)
{
    PlanNode child = aggregation.getSource();

    Optional<FilterNode> filter = findFirstPlanNodeMatching(
            child,
            FilterNode.class::isInstance,
            Predicates.<PlanNode>isInstance(ProjectNode.class).or(LimitNode.class::isInstance));

    if (!filter.isPresent()) {
        return Optional.of(new Result(ImmutableList.of(), child));
    }

    Expression predicate = filter.get().getPredicate();

    if (!isSupportedPredicate(predicate)) {
        return Optional.empty();
    }

    if (!DependencyExtractor.extractAll(predicate).containsAll(boundVariables)) {
        return Optional.empty();
    }

    ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
    ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();

    for (Expression conjunct : extractConjuncts(predicate)) {
        List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);

        if (boundVariables.stream().anyMatch(symbols::contains)) {
            joinExpressions.add(conjunct);
        }
        else {
            newPredicate.add(conjunct);
        }
    }
    ...
}
```

File: presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java]: I am going to store taskId in 16 bits, stageId 16 bits and counter in 32 bits, in case of overflow any of these a runtime exception will be thrown

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/EnforceUniqueColumnsOperator.java]: I am going to leverage taskId and stageId to make sure that rows from two different stages for `AssigneUniqueId` are unique.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java (hunk 8 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/EnforceUniqueColumns.java]: 8-10 bits for stage I'd should be more than sufficient. It's rare to see a well-behaved query with more than a few dozen stages.

Task ids are per stage, and there's one per machine, so 12-14 bits should cover any reasonable cluster sizes for the foreseeable future.

32 bits for row id are probably not sufficient. We've seen tasks that generate more than that many rows. I'd use more bits for that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: done, done ;)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java (hunk 11 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java (hunk 17 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java (hunk 18 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: done

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationSubqueryToLeftOuterJoin.java]: Agree, that way code looks much more readable. Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java (hunk 12 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: Yes. Done.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: It is kind of separate changes. I will extract them to separate commits.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java (hunk 13 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: I like the builder idea.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedScalarToJoin.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoApplyNodeLeftChecker.java (hunk 14 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoApplyNodeLeftChecker.java]: done

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 15 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: done. I changed the column from decimal (float) to date.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 16 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: done

**Comment by kokosing** [comment]
@martint All the comments applied and superseded by #6019
